java/src/main/native/src/TableJni.cpp

```diff
@@ -3044,13 +3044,20 @@ JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_Table_dropDuplicates(
     auto const keys_indices =
         std::vector<cudf::size_type>(native_keys_indices.begin(), native_keys_indices.end());
 
-    auto result = cudf::drop_duplicates(
-        *input, keys_indices,
-        keep_first ? cudf::duplicate_keep_option::KEEP_FIRST :
-                     cudf::duplicate_keep_option::KEEP_LAST,
-        nulls_equal ? cudf::null_equality::EQUAL : cudf::null_equality::UNEQUAL,
-        nulls_before ? cudf::null_order::BEFORE : cudf::null_order::AFTER,
-        rmm::mr::get_current_device_resource());
+    // cudf::drop_duplicates works like std::unique thus does NOT match the behavior of
+    // pandas.DataFrame.drop_duplicates. Users need to stable sort the input first and then drop.
+    std::vector<cudf::order> order(keys_indices.size(), cudf::order::ASCENDING);
+    std::vector<cudf::null_order> null_precedence(
+        keys_indices.size(), nulls_before ? cudf::null_order::BEFORE : cudf::null_order::AFTER);
+    auto const sorted_input =
+        cudf::stable_sort_by_key(*input, input->select(keys_indices), order, null_precedence);
```

Comment:
That this change preserves the existing semantics of `Java_ai_rapids_cudf_Table_dropDuplicates()` is deeply appreciated. Thank you!

I had to consult @jlowe to confirm. It appears that `dropDuplicates()` is not invoked from the `spark-rapids` code at least, given the cost of the (re-)sorting. `spark-rapids` might eventually prefer to use the better performing `unordered_drop_duplicates()`, which currently has no JNI bindings.

I'm slightly concerned now that with this change, there are semantic differences between `cudf::drop_duplicates()` and its JNI binding. The former drops adjacent duplicates, and the latter removes all duplicates. It might be less confusing to the eventual user to keep these aligned.

One option might be to rename `cudf::drop_duplicates()` to `cudf::drop_adjacent_duplicates()`, but keep the JNI function `Java_ai_rapids_cudf_Table_dropDuplicates()`. Different functions with different semantics.

Reply:
This looks good.